File: Algorithms/Searching & Sorting/Heap Sort/heap_sort.py
def heapify(arr, n, i):
    """This is to max heapify
        Parameters:
            arr(list): List of integers
            n(int): Length of list
            i(int): Root element
    """
    largest = i  
    left = 2 * i + 1     
    right = 2 * i + 2     
   
    if left < n and arr[i] < arr[left]: 
        largest = left 
  
    if right < n and arr[largest] < arr[right]: 
        largest = right 
  
    if largest != i: 
        arr[i],arr[largest] = arr[largest],arr[i] # swap 
        heapify(arr, n, largest) 

 
def heapSort(arr):
    """This function sort the integers.
        Parameters:
            arr(list): List of integers
        Returns:
            None
    """
    n = len(arr) 
  
    # Build a maxheap. 
    for i in range(n, -1, -1): 
        heapify(arr, n, i) 
  
    # One by one extract elements 
    for i in range(n-1, 0, -1): 
        arr[i], arr[0] = arr[0], arr[i] # swap 
        heapify(arr, i, 0) 


def main():
    arr = [ 12, 11, 13, 5, 6, 7] 
    heapSort(arr) 
    print("Sorted array is",arr)

    
if __name__ == "__main__":
    main()

File: Algorithms/Searching & Sorting/Heap Sort/HeapSort.java
public class HeapSort {

	private int n;

	public HeapSort() {
	}

	public int[] sort(int[] array) {
		buildHeap(array);

		for (int i = n; i > 0; i--) {
			swap(array, 0, i);
			n -= 1;
			maxHeap(array, 0);
		}

		return array;
	}

	private void buildHeap(int[] a) {
		n = a.length - 1;
		for (int i = n / 2; i >= 0; i--) {
			maxHeap(a, i);
		}
	}

	private void maxHeap(int[] a, int i) {
		int left = 2 * i;
		int right = 2 * i + 1;
		int largest = 0;

		if (left <= n && a[left] > a[i]) {
			largest = left;
		} else {
			largest = i;
		}
		
		if (right <= n && a[right] > a[largest]) {
			largest = right;
		} 
		
		if (largest != i) {
			swap(a, i, largest);
			maxHeap(a, largest);
		}
	}

	private void swap(int a[], int i, int j) {
		int tmp = a[i];
		a[i] = a[j];
		a[j] = tmp;
	}
	
	public static void main (String args[]) {
		int testArray[] = {3, 5, 1, 0, 9, 7, 4, 8, 2, 6};
		
		HeapSort heapSort = new HeapSort();
		int sortedArray[] = heapSort.sort(testArray);
		
		for (int i=0; i < sortedArray.length; i++) {
			System.out.print(sortedArray[i] + " ");
		}
	}

}
